drivers/gpio/gpio_aw9523b.c

```diff
@@ -0,0 +1,504 @@
+}
+
+static int gpio_aw9523b_port_set_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	return gpio_aw9523b_port_read_write_toggle(dev, pins, &pins, WRITE);
+}
+
+static int gpio_aw9523b_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	const gpio_port_value_t zero = 0;
+
+	/* If WRITE is specified, this pointer is used for reading only. It can cast. */
+	return gpio_aw9523b_port_read_write_toggle(dev, pins, (gpio_port_value_t *)&zero, WRITE);
+}
+
+static int gpio_aw9523b_port_toggle_bits(const struct device *dev, gpio_port_pins_t pins)
+{
+	return gpio_aw9523b_port_read_write_toggle(dev, pins, NULL, TOGGLE);
+}
+
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
+static void gpio_aw9523b_interrupt_worker(struct k_work *work)
+{
+	struct gpio_aw9523b_data *const data =
+		CONTAINER_OF(work, struct gpio_aw9523b_data, intr_worker);
+	const struct gpio_aw9523b_config *const config = data->dev->config;
+	gpio_port_value_t value, rising, falling;
+	uint8_t buf[2];
+	int err;
+
+	/*
+	 * We need to read INPUT0 to deassert INTN when that is asserted by
+	 * pin0-7 interruption, and same also INPUT1 for pin8-15.
+	 * It cannot deassert by burst-read.
+	 */
+	err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, &buf[0], 1);
+	if (err) {
+		LOG_ERR("%s: Failed to read INPUT0 %d", data->dev->name, err);
+	}
+	err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT1, &buf[1], 1);
+	if (err) {
+		LOG_ERR("%s: Failed to read INPUT1 %d", data->dev->name, err);
+	}
+
+	value = sys_get_le16(buf);
+
+	rising = (value ^ data->prev_value) & (value & data->rising_event_pins);
+	falling = (value ^ data->prev_value) & (~value & data->falling_event_pins);
+
+	data->prev_value = value;
+
+	gpio_fire_callbacks(&data->callbacks, data->dev, rising | falling);
+}
+
+static int gpio_aw9523b_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+						enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	const struct gpio_aw9523b_config *const config = dev->config;
+	struct gpio_aw9523b_data *const data = dev->data;
+	const uint8_t port = (pin < 8) ? 0 : 1;
+	const uint8_t mask = BIT(port ? pin - 8 : pin);
+	const uint8_t n_int_en = (mode & GPIO_INT_MODE_EDGE) ? 0x00 : 0xFF;
+	uint8_t buf[2];
+	int err;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	if (mode == GPIO_INT_MODE_LEVEL) {
+		return -ENOTSUP;
+	}
+	if (data->common.invert & BIT(pin)) {
+		WRITE_BIT(data->falling_event_pins, pin, trig & GPIO_INT_HIGH_1);
+		WRITE_BIT(data->rising_event_pins, pin, trig & GPIO_INT_LOW_0);
+	} else {
+		WRITE_BIT(data->falling_event_pins, pin, trig & GPIO_INT_LOW_0);
+		WRITE_BIT(data->rising_event_pins, pin, trig & GPIO_INT_HIGH_1);
+	}
+
+	k_sem_take(aw9523b_get_lock(config->mfd_dev), K_FOREVER);
+
+	err = i2c_reg_update_byte_dt(&config->i2c, AW9523B_REG_INT(port), mask, n_int_en);
+	if (err) {
+		LOG_ERR("%s: Failed to configure pin interruption (%d)", dev->name, err);
```

Comment:
you sure you want to go a head and keep running the rest of the code if this fails? seems a bit pointless, the bus is likely gone for good if we get here, and even if the following call works, what good does it do for us? may as well `return err` and `return 0` at the end

Reply:
As you pointed out, there was a route just below this that could cause the semaphore release to leak, so we addressed that as well.